validator/accounts/status.go

```diff
@@ -0,0 +1,176 @@
+	defer span.End()
+
+	errorChannel := make(chan error, MaxRequestLimit)
+	statusChannel := make(chan []ValidatorStatusMetadata, MaxRequestLimit)
+	// Launch routines to fetch statuses.
+	i, numBatches := 0, 0
+	for ; i+MaxRequestKeys < len(pubkeys); i += MaxRequestKeys {
+		go fetchValidatorStatus(
+			ctx, beaconNodeRPCProvider, pubkeys[i:i+MaxRequestKeys], statusChannel, errorChannel)
+		numBatches++
+	}
+	if i < len(pubkeys) {
+		go fetchValidatorStatus(
+			ctx, beaconNodeRPCProvider, pubkeys[i:], statusChannel, errorChannel)
+		numBatches++
+	}
+	// Wait for fetch routines to finish.
+	var err error
+	allStatuses := make([][]ValidatorStatusMetadata, 0, numBatches)
+	for i := 0; i < numBatches; i++ {
+		select {
+		case statuses := <-statusChannel:
+			allStatuses = append(allStatuses, statuses)
+		case e := <-errorChannel:
+			err = e
+		}
+	}
+
+	return allStatuses, err
+}
+
+func fetchValidatorStatus(
+	ctx context.Context,
+	rpcProvder ethpb.BeaconNodeValidatorClient,
+	pubkeys [][]byte,
+	statusChannel chan []ValidatorStatusMetadata,
+	errorChannel chan error) {
+	if ctx.Err() == context.Canceled {
+		errorChannel <- errors.Wrap(ctx.Err(), "context has been canceled.")
+		return
+	}
+
+	req := &ethpb.MultipleValidatorStatusRequest{PublicKeys: pubkeys}
+	resp, err := rpcProvder.MultipleValidatorStatus(ctx, req)
+	if err != nil {
+		errorChannel <- errors.Wrap(
+			err,
+			fmt.Sprintf("Failed to fetch validator statuses for %d key(s)", len(pubkeys)))
+		return
+	}
+
+	statusChannel <- responseToSortedMetadata(resp)
+}
+
+func responseToSortedMetadata(resp *ethpb.MultipleValidatorStatusResponse) []ValidatorStatusMetadata {
+	pubkeys := resp.GetPublicKeys()
+	validatorStatuses := make([]ValidatorStatusMetadata, len(pubkeys))
+	for i, status := range resp.GetStatuses() {
+		validatorStatuses[i] = ValidatorStatusMetadata{
+			PublicKey: pubkeys[i],
+			Metadata:  status,
+		}
+	}
+	sort.Slice(validatorStatuses, func(i, j int) bool {
+		return validatorStatuses[i].Metadata.Status < validatorStatuses[j].Metadata.Status
+	})
+	return validatorStatuses
+}
+
+// ExtractPublicKeys extracts only the public keys from the decrypted keys from the keystore.
+func ExtractPublicKeys(decryptedKeys map[string]*keystore.Key) [][]byte {
+	i := 0
+	pubkeys := make([][]byte, len(decryptedKeys))
+	for _, key := range decryptedKeys {
+		pubkeys[i] = key.PublicKey.Marshal()
+		i++
+	}
+	return pubkeys
+}
+
+// MergeStatuses merges k sorted ValidatorStatusMetadata slices to 1.
+func MergeStatuses(allStatuses [][]ValidatorStatusMetadata) []ValidatorStatusMetadata {
+	if len(allStatuses) == 0 {
+		return []ValidatorStatusMetadata{}
+	}
+	if len(allStatuses) == 1 {
+		return allStatuses[0]
+	}
+	leftHalf := allStatuses[:len(allStatuses)/2]
+	rightHalf := allStatuses[len(allStatuses)/2:]
+	return mergeTwo(MergeStatuses(leftHalf), MergeStatuses(rightHalf))
+}
+
+func mergeTwo(s1, s2 []ValidatorStatusMetadata) []ValidatorStatusMetadata {
```

Comment:
Why do we need all this logic? Can't we have a simple for loop in Go to flatten the slice:

```go
for i := 0; i < len(items); i++ {
    statuses = append(statuses, items[i]...)
}
```

Reply:
https://github.com/prysmaticlabs/prysm/pull/5784/commits/93793bebd2994e1ea71b81f47afe89b808dc3f0f